kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/ModalContainer.kt

```diff
@@ -0,0 +1,204 @@
+
+    // Create a new body view each time the type of [AlertContainerScreen.base] changes.
+    takeScreensSubs.add(screens
+        .whileAttached()
+        .distinctUntilChanged { containerScreen -> containerScreen.baseScreen::class }
+        .subscribe {
+          removeAllViews()
+          addView(it.viewForBase(screens, viewRegistry, this))
+        }
+    )
+
+    // Compare the new modals list to the set of dialogs already on display. Show
+    // any new ones, throw away any stale ones.
+    takeScreensSubs.add(screens
+        .whileAttached()
+        .subscribe { containerScreen ->
+          val newDialogs = mutableListOf<DialogRef<M>>()
+          for ((i, modal) in containerScreen.modals.withIndex()) {
+            newDialogs += if (i < dialogs.size && dialogs[i].screen.matches(modal)) {
+              dialogs[i]
+            } else {
+              val modalsInThisLayer = screens.whileAttached()
+                  .filter { i < it.modals.size }
+                  .map { it.modals[i] }
+              DialogRef(modal, containerScreen.showDialog(i, modalsInThisLayer, viewRegistry))
+            }
+          }
+
+          (dialogs - newDialogs).forEach { it.dialog.hide() }
+          dialogs = newDialogs
+        }
+    )
+  }
+
+  /**
+   * Returns true if the new screen can be shown by the dialog that was created for the receiver.
+   * Default implementation compares equality of the receiver and the new screen.
+   */
+  protected open fun M.matches(nextModal: M): Boolean = this == nextModal
+
+  protected abstract fun showDialog(
+    modalScreen: M,
+    screens: Observable<out M>,
+    viewRegistry: ViewRegistry
+  ): Dialog
+
+  private fun HasModals<*, M>.showDialog(
+    index: Int,
+    screens: Observable<out M>,
+    viewRegistry: ViewRegistry
+  ): Dialog = showDialog(modals[index], screens, viewRegistry)
+
+  private fun <T> Observable<T>.whileAttached(): Observable<T> =
+    attached.switchMap { isAttached -> if (isAttached) this else Observable.never<T>() }
+
+  protected class DialogRef<M>(
+    val screen: M,
+    val dialog: Dialog
+  )
+
+  private fun <T : Any> HasModals<T, *>.viewForBase(
+    containerScreens: Observable<out HasModals<*, *>>,
+    viewRegistry: ViewRegistry,
+    container: ViewGroup
+  ): View {
+    val baseScreens: Observable<out T> = containerScreens.mapToBaseMatching(this)
+    val binding: ViewBinding<T> = viewRegistry.getBinding(baseScreen::class.jvmName)
+    return binding.buildView(baseScreens, viewRegistry, container)
+  }
+
+  private fun <T : Any> Observable<out HasModals<*, *>>.mapToBaseMatching(
+    screen: HasModals<T, *>
+  ): Observable<out T> = map { it.baseScreen }.ofType(screen.baseScreen::class.java)
+
+  protected fun <M : Any> HasModals<*, M>.viewForModal(
+    containerScreens: Observable<out HasModals<*, *>>,
+    index: Int,
+    viewRegistry: ViewRegistry,
+    container: ViewGroup
+  ): View {
+    if (index < modals.size) {
```

Comment:
If index is less than modals.size, we throw an exception saying that index must be less than modals.size? What am I missing?

Reply:
Deleted this whole function, it was never called.